<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>encoded.commands.generate_annotations &#8212; 4D Nucleome Portal 1.0 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../_static/sidebar.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">4D Nucleome Portal 1.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../../encoded.html" accesskey="U">encoded</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for encoded.commands.generate_annotations</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">import</span> <span class="nn">json</span>

<span class="n">EPILOG</span> <span class="o">=</span> <span class="vm">__doc__</span>

<span class="n">_HGNC_FILE</span> <span class="o">=</span> <span class="s1">&#39;https://s3-us-west-1.amazonaws.com/encoded-build/annotations/human/hgnc_dump_03-30-2015.tsv&#39;</span>
<span class="n">_MOUSE_FILE</span> <span class="o">=</span> <span class="s1">&#39;https://s3-us-west-1.amazonaws.com/encoded-build/annotations/mouse/mouse.txt&#39;</span>
<span class="n">_DM_FILE</span> <span class="o">=</span> <span class="s1">&#39;https://s3-us-west-1.amazonaws.com/encoded-build/annotations/drosophila/dm.txt&#39;</span>
<span class="n">_CE_FILE</span> <span class="o">=</span> <span class="s1">&#39;https://s3-us-west-1.amazonaws.com/encoded-build/annotations/worm/c_elegans.txt&#39;</span>
<span class="n">_ENSEMBL_URL</span> <span class="o">=</span> <span class="s1">&#39;http://rest.ensembl.org/&#39;</span>
<span class="n">_GENEINFO_URL</span> <span class="o">=</span> <span class="s1">&#39;http://mygene.info/v2/gene/&#39;</span>


<div class="viewcode-block" id="get_annotation"><a class="viewcode-back" href="../../../pydoc/encoded.commands.generate_annotations.html#encoded.commands.generate_annotations.get_annotation">[docs]</a><span class="k">def</span> <span class="nf">get_annotation</span><span class="p">():</span>
    <span class="k">return</span> <span class="p">{</span>
        <span class="s1">&#39;assembly_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="s1">&#39;chromosome&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span>
    <span class="p">}</span></div>


<div class="viewcode-block" id="assembly_mapper"><a class="viewcode-back" href="../../../pydoc/encoded.commands.generate_annotations.html#encoded.commands.generate_annotations.assembly_mapper">[docs]</a><span class="k">def</span> <span class="nf">assembly_mapper</span><span class="p">(</span><span class="n">location</span><span class="p">,</span> <span class="n">species</span><span class="p">,</span> <span class="n">input_assembly</span><span class="p">,</span> <span class="n">output_assembly</span><span class="p">):</span>
    <span class="c1"># All others</span>
    <span class="n">new_url</span> <span class="o">=</span> <span class="n">_ENSEMBL_URL</span> <span class="o">+</span> <span class="s1">&#39;map/&#39;</span> <span class="o">+</span> <span class="n">species</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> \
        <span class="o">+</span> <span class="n">input_assembly</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">location</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">output_assembly</span> \
        <span class="o">+</span> <span class="s1">&#39;/?content-type=application/json&#39;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">new_response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">new_url</span><span class="p">)</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">return</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_response</span><span class="p">[</span><span class="s1">&#39;mappings&#39;</span><span class="p">]):</span>
            <span class="k">return</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">new_response</span><span class="p">[</span><span class="s1">&#39;mappings&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;mapped&#39;</span><span class="p">]</span>
        <span class="n">chromosome</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;seq_region_name&#39;</span><span class="p">]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span>
        <span class="k">return</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span></div>


<div class="viewcode-block" id="human_annotations"><a class="viewcode-back" href="../../../pydoc/encoded.commands.generate_annotations.html#encoded.commands.generate_annotations.human_annotations">[docs]</a><span class="k">def</span> <span class="nf">human_annotations</span><span class="p">(</span><span class="n">es</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates JSON from TSV files</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">species</span> <span class="o">=</span> <span class="s1">&#39; (homo sapiens)&#39;</span>
    <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">_HGNC_FILE</span><span class="p">)</span>
    <span class="n">header</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">response</span><span class="o">.</span><span class="n">content</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
        <span class="c1"># Populating headers and skipping header row</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">header</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">header</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="n">r</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)))</span>

        <span class="c1"># Ensembl ID is used to grab annotations for different references</span>
        <span class="k">if</span> <span class="s1">&#39;Ensembl Gene ID&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">r</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">elif</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Ensembl Gene ID&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">continue</span>
            
        <span class="c1"># Annotations are keyed by Gene ID in ES</span>
        <span class="k">if</span> <span class="s1">&#39;Entrez Gene ID&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">r</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="c1"># Assumption: payload.id and id should always be same</span>
        <span class="n">doc</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;annotations&#39;</span><span class="p">:</span> <span class="p">[]}</span>
        <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;name_suggest&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;input&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;Approved Name&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">species</span><span class="p">,</span>
                      <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Approved Symbol&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">species</span><span class="p">,</span>
                      <span class="n">r</span><span class="p">[</span><span class="s1">&#39;HGNC ID&#39;</span><span class="p">],</span>
                      <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Entrez Gene ID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; (Gene ID)&#39;</span><span class="p">],</span>
            <span class="s1">&#39;payload&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;HGNC ID&#39;</span><span class="p">],</span>
                        <span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="n">species</span><span class="p">}</span>
        <span class="p">}</span>
        <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;HGNC ID&#39;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Entrez Gene ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isdigit</span><span class="p">():</span>
            <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Entrez Gene ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;Entrez Gene ID&#39;</span><span class="p">])</span>

        <span class="c1"># Adding gene synonyms to autocomplete</span>
        <span class="k">if</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="n">synonyms</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="o">+</span> <span class="n">species</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)]</span>
            <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;name_suggest&#39;</span><span class="p">][</span><span class="s1">&#39;input&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;name_suggest&#39;</span><span class="p">][</span><span class="s1">&#39;input&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">synonyms</span>

        <span class="n">url</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{ensembl}</span><span class="s1">lookup/id/</span><span class="si">{id}</span><span class="s1">?content-type=application/json&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">ensembl</span><span class="o">=</span><span class="n">_ENSEMBL_URL</span><span class="p">,</span>
            <span class="nb">id</span><span class="o">=</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;Ensembl Gene ID&#39;</span><span class="p">])</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">annotation</span> <span class="o">=</span> <span class="n">get_annotation</span><span class="p">()</span>
            <span class="k">if</span> <span class="s1">&#39;assembly_name&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">response</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;assembly_name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;assembly_name&#39;</span><span class="p">]</span>
            <span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;chromosome&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;seq_region_name&#39;</span><span class="p">]</span>
            <span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>
            <span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span>
            <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;annotations&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">annotation</span><span class="p">)</span>

            <span class="c1"># Get GRcH37 annotation</span>
            <span class="n">location</span> <span class="o">=</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;seq_region_name&#39;</span><span class="p">]</span> \
                <span class="o">+</span> <span class="s1">&#39;:&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">response</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">])</span> \
                <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">response</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">])</span>
            <span class="n">ann</span> <span class="o">=</span> <span class="n">get_annotation</span><span class="p">()</span>
            <span class="n">ann</span><span class="p">[</span><span class="s1">&#39;assembly_name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;GRCh37&#39;</span>
            <span class="n">ann</span><span class="p">[</span><span class="s1">&#39;chromosome&#39;</span><span class="p">],</span> <span class="n">ann</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">ann</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">=</span> \
                <span class="n">assembly_mapper</span><span class="p">(</span><span class="n">location</span><span class="p">,</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;species&#39;</span><span class="p">],</span>
                                <span class="s1">&#39;GRCh38&#39;</span><span class="p">,</span> <span class="s1">&#39;GRCh37&#39;</span><span class="p">)</span>
            <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;annotations&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ann</span><span class="p">)</span>
        <span class="n">annotations</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;_index&quot;</span><span class="p">:</span> <span class="s2">&quot;annotations&quot;</span><span class="p">,</span>
                <span class="s2">&quot;_type&quot;</span><span class="p">:</span> <span class="s2">&quot;default&quot;</span><span class="p">,</span>
                <span class="s2">&quot;_id&quot;</span><span class="p">:</span> <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span>
            <span class="p">}</span>
        <span class="p">})</span>
        <span class="n">annotations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">doc</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">annotations</span></div>


<div class="viewcode-block" id="mouse_annotations"><a class="viewcode-back" href="../../../pydoc/encoded.commands.generate_annotations.html#encoded.commands.generate_annotations.mouse_annotations">[docs]</a><span class="k">def</span> <span class="nf">mouse_annotations</span><span class="p">(</span><span class="n">mouse_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Updates and get JSON file for mouse annotations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mouse_file</span><span class="p">)</span>
    <span class="n">header</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">response</span><span class="o">.</span><span class="n">content</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
        <span class="c1"># skipping header row</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">header</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">header</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="n">r</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)))</span>
        <span class="k">if</span> <span class="s1">&#39;Chromosome Name&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">r</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">doc</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;annotations&#39;</span><span class="p">:</span> <span class="p">[]}</span>
        <span class="n">species</span> <span class="o">=</span> <span class="s1">&#39; (mus musculus)&#39;</span>
        <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;name_suggest&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;input&#39;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s1">&#39;payload&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Ensembl Gene ID&#39;</span><span class="p">],</span>
                        <span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="n">species</span><span class="p">}</span>
        <span class="p">}</span>
        <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Ensembl Gene ID&#39;</span><span class="p">]</span>

        <span class="k">if</span> <span class="s1">&#39;MGI symbol&#39;</span> <span class="ow">in</span> <span class="n">r</span> <span class="ow">and</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;MGI symbol&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;name_suggest&#39;</span><span class="p">][</span><span class="s1">&#39;input&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;MGI symbol&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">species</span><span class="p">)</span>

        <span class="k">if</span> <span class="s1">&#39;MGI ID&#39;</span> <span class="ow">in</span> <span class="n">r</span> <span class="ow">and</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;MGI ID&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;name_suggest&#39;</span><span class="p">][</span><span class="s1">&#39;input&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;MGI ID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">species</span><span class="p">)</span>

        <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;annotations&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s1">&#39;assembly_name&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCm38&#39;</span><span class="p">,</span>
            <span class="s1">&#39;chromosome&#39;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Chromosome Name&#39;</span><span class="p">],</span>
            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Gene Start (bp)&#39;</span><span class="p">],</span>
            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Gene End (bp)&#39;</span><span class="p">]</span>
        <span class="p">})</span>

        <span class="n">mm9_url</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{geneinfo}{ensembl}</span><span class="s1">?fields=genomic_pos_mm9&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">geneinfo</span><span class="o">=</span><span class="n">_GENEINFO_URL</span><span class="p">,</span>
            <span class="n">ensembl</span><span class="o">=</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;Ensembl Gene ID&#39;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mm9_url</span><span class="p">)</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="s1">&#39;genomic_pos_mm9&#39;</span> <span class="ow">in</span> <span class="n">response</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">response</span><span class="p">[</span><span class="s1">&#39;genomic_pos_mm9&#39;</span><span class="p">],</span> <span class="nb">dict</span><span class="p">):</span>
                    <span class="n">ann</span> <span class="o">=</span> <span class="n">get_annotation</span><span class="p">()</span>
                    <span class="n">ann</span><span class="p">[</span><span class="s1">&#39;assembly_name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;GRCm37&#39;</span>
                    <span class="n">ann</span><span class="p">[</span><span class="s1">&#39;chromosome&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;genomic_pos_mm9&#39;</span><span class="p">][</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span>
                    <span class="n">ann</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;genomic_pos_mm9&#39;</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>
                    <span class="n">ann</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;genomic_pos_mm9&#39;</span><span class="p">][</span><span class="s1">&#39;end&#39;</span><span class="p">]</span>
                    <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;annotations&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ann</span><span class="p">)</span>
        <span class="n">annotations</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;_index&quot;</span><span class="p">:</span> <span class="s2">&quot;annotations&quot;</span><span class="p">,</span>
                <span class="s2">&quot;_type&quot;</span><span class="p">:</span> <span class="s2">&quot;default&quot;</span><span class="p">,</span>
                <span class="s2">&quot;_id&quot;</span><span class="p">:</span> <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span>
            <span class="p">}</span>
        <span class="p">})</span>
        <span class="n">annotations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">doc</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">annotations</span></div>


<div class="viewcode-block" id="other_annotations"><a class="viewcode-back" href="../../../pydoc/encoded.commands.generate_annotations.html#encoded.commands.generate_annotations.other_annotations">[docs]</a><span class="k">def</span> <span class="nf">other_annotations</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">species</span><span class="p">,</span> <span class="n">assembly</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates C. elegans and drosophila annotaions</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
    <span class="n">header</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">response</span><span class="o">.</span><span class="n">content</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
        <span class="c1"># skipping header row</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">header</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">header</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="n">r</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)))</span>
        <span class="k">if</span> <span class="s1">&#39;Chromosome Name&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">r</span> <span class="ow">or</span> <span class="s1">&#39;Ensembl Gene ID&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">r</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">doc</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;annotations&#39;</span><span class="p">:</span> <span class="p">[]}</span>
        <span class="n">annotation</span> <span class="o">=</span> <span class="n">get_annotation</span><span class="p">()</span>

        <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;name_suggest&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;input&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;Associated Gene Name&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">species</span><span class="p">],</span>
            <span class="s1">&#39;payload&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Ensembl Gene ID&#39;</span><span class="p">],</span>
                        <span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="n">species</span><span class="p">}</span>
        <span class="p">}</span>
        <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Ensembl Gene ID&#39;</span><span class="p">]</span>
        <span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;assembly_name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">assembly</span>
        <span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;chromosome&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Chromosome Name&#39;</span><span class="p">]</span>
        <span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Gene Start (bp)&#39;</span><span class="p">]</span>
        <span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;Gene End (bp)&#39;</span><span class="p">]</span>
        <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;annotations&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">annotation</span><span class="p">)</span>
        <span class="n">annotations</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;_index&quot;</span><span class="p">:</span> <span class="s2">&quot;annotations&quot;</span><span class="p">,</span>
                <span class="s2">&quot;_type&quot;</span><span class="p">:</span> <span class="s2">&quot;default&quot;</span><span class="p">,</span>
                <span class="s2">&quot;_id&quot;</span><span class="p">:</span> <span class="n">doc</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span>
            <span class="p">}</span>
        <span class="p">})</span>
        <span class="n">annotations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">doc</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">annotations</span></div>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../pydoc/encoded.commands.generate_annotations.html#encoded.commands.generate_annotations.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Get annotations from multiple sources</span>
<span class="sd">    This helps to implement autocomplete for region search</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="kn">import</span> <span class="nn">argparse</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Generate annotations JSON file for multiple species&quot;</span><span class="p">,</span>
        <span class="n">epilog</span><span class="o">=</span><span class="n">EPILOG</span><span class="p">,</span> <span class="n">formatter_class</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">RawDescriptionHelpFormatter</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="n">other_annotations</span><span class="p">(</span><span class="n">_DM_FILE</span><span class="p">,</span> <span class="s1">&#39; (D. melanogaster)&#39;</span><span class="p">,</span> <span class="s1">&#39;BDGP6&#39;</span><span class="p">)</span> <span class="o">+</span>\
        <span class="n">other_annotations</span><span class="p">(</span><span class="n">_CE_FILE</span><span class="p">,</span> <span class="s1">&#39; (C. elegans)&#39;</span><span class="p">,</span> <span class="s1">&#39;WBcel235&#39;</span><span class="p">)</span> <span class="o">+</span>\
        <span class="n">human_annotations</span><span class="p">(</span><span class="n">_HGNC_FILE</span><span class="p">)</span> <span class="o">+</span>\
        <span class="n">mouse_annotations</span><span class="p">(</span><span class="n">_MOUSE_FILE</span><span class="p">)</span>

    <span class="c1"># Create annotations JSON file</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;annotations.json&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
        <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">outfile</span><span class="p">)</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">4D Nucleome Portal 1.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../../encoded.html" >encoded</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, 4DN-DCIC.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.3.
    </div>
  </body>
</html>