<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>encoded.types.experiment module &#8212; 4D Nucleome Portal 1.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/sidebar.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="encoded.types.experiment_set module" href="encoded.types.experiment_set.html" />
    <link rel="prev" title="encoded.types.biosource module" href="encoded.types.biosource.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="encoded.types.experiment_set.html" title="encoded.types.experiment_set module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="encoded.types.biosource.html" title="encoded.types.biosource module"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">4D Nucleome Portal 1.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="modules.html" >Python API Reference</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="encoded.html" >encoded package</a> &#187;</li>
          <li class="nav-item nav-item-3"><a href="encoded.types.html" accesskey="U">encoded.types package</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-encoded.types.experiment">
<span id="encoded-types-experiment-module"></span><h1>encoded.types.experiment module<a class="headerlink" href="#module-encoded.types.experiment" title="Permalink to this headline">¶</a></h1>
<p>Abstract collection for experiment and integration of all experiment types.</p>
<dl class="class">
<dt id="encoded.types.experiment.Experiment">
<em class="property">class </em><code class="descclassname">encoded.types.experiment.</code><code class="descname">Experiment</code><span class="sig-paren">(</span><em>registry</em>, <em>models</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#Experiment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.Experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="encoded.types.base.html#encoded.types.base.Item" title="encoded.types.base.Item"><code class="xref py py-class docutils literal"><span class="pre">encoded.types.base.Item</span></code></a></p>
<p>The main experiment class.</p>
<dl class="attribute">
<dt id="encoded.types.experiment.Experiment.base_types">
<code class="descname">base_types</code><em class="property"> = ['Experiment', 'Item']</em><a class="headerlink" href="#encoded.types.experiment.Experiment.base_types" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.Experiment.embedded">
<code class="descname">embedded</code><em class="property"> = ['protocol', 'protocol_variation', 'lab', 'award', 'biosample', 'biosample.biosource', 'biosample.modifications', 'biosample.treatments', 'biosample.biosource.individual.organism']</em><a class="headerlink" href="#encoded.types.experiment.Experiment.embedded" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="encoded.types.experiment.Experiment.experiment_sets">
<code class="descname">experiment_sets</code><span class="sig-paren">(</span><em>request</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#Experiment.experiment_sets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.Experiment.experiment_sets" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="encoded.types.experiment.Experiment.find_current_sop_map">
<code class="descname">find_current_sop_map</code><span class="sig-paren">(</span><em>experiment_type</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#Experiment.find_current_sop_map"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.Experiment.find_current_sop_map" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="encoded.types.experiment.Experiment.generate_mapid">
<code class="descname">generate_mapid</code><span class="sig-paren">(</span><em>experiment_type</em>, <em>num</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#Experiment.generate_mapid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.Experiment.generate_mapid" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.Experiment.name_key">
<code class="descname">name_key</code><em class="property"> = 'accession'</em><a class="headerlink" href="#encoded.types.experiment.Experiment.name_key" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.Experiment.schema">
<code class="descname">schema</code><em class="property"> = OrderedDict([('title', 'Experiment'), ('description', 'Schema for submitting metadata for an assay.'), ('id', '/profiles/experiment.json'), ('$schema', 'http://json-schema.org/draft-04/schema#'), ('type', 'object'), ('required', ['experiment_type', 'biosample', 'award', 'lab']), ('identifyingProperties', ['uuid', 'accession', 'aliases']), ('additionalProperties', False), ('dependencies', OrderedDict([('biosample_quantity', ['biosample_quantity_units']), ('biosample_quantity_units', ['biosample_quantity'])])), ('properties', OrderedDict([('status', OrderedDict([('title', 'Status'), ('type', 'string'), ('default', 'in review by lab'), ('permission', 'import_items'), ('enum', ['released', 'current', 'revoked', 'deleted', 'replaced', 'in review by lab', 'in review by project', 'released to project'])])), ('accession', OrderedDict([('accessionType', 'EX')])), ('description', OrderedDict([('title', 'description'), ('description', 'A short description of the experiment'), ('type', 'string')])), ('experiment_type', OrderedDict([('type', 'string'), ('description', 'A controlled term specifying the type of experiment.'), ('internal_comment', 'should be ontology term?')])), ('protocol', OrderedDict([('title', 'Reference Protocol'), ('description', 'Reference Protocol Document.'), ('type', 'string'), ('linkTo', 'Protocol')])), ('protocol_variation', OrderedDict([('title', 'Protocol Documents'), ('description', 'Protocol Documents Describing the variations in the protocol from reference.'), ('type', 'array'), ('items', OrderedDict([('title', 'Document'), ('description', 'A document associated with the specific experimental protocol.'), ('type', 'string'), ('linkTo', 'Protocol')]))])), ('experiment_relation', OrderedDict([('title', 'Experiment relations'), ('description', 'All related experiments'), ('type', 'array'), ('items', OrderedDict([('title', 'Experiment relation'), ('type', 'object'), ('additionalProperties', False), ('properties', OrderedDict([('relationship_type', OrderedDict([('type', 'string'), ('description', 'A controlled term specifying the relationship between experiments.'), ('title', 'Relationship Type'), ('enum', ['controlled by', 'control for', 'derived from', 'source for'])])), ('experiment', OrderedDict([('type', 'string'), ('description', 'The related experiment'), ('linkTo', 'Experiment')]))]))]))])), ('biosample', OrderedDict([('title', 'Biological Sample'), ('description', 'Starting biological material.'), ('type', 'string'), ('linkTo', 'Biosample')])), ('biosample_quantity', OrderedDict([('title', 'Biological Sample Quantity'), ('description', 'The amount of starting Biological sample going into the experiment'), ('type', 'number')])), ('biosample_quantity_units', OrderedDict([('title', 'Quantity Units'), ('description', 'The units that go along with the biological sample quantity'), ('type', 'string'), ('enum', ['g', 'mg', 'μg', 'ml', 'cells'])])), ('files', OrderedDict([('title', 'Data files'), ('description', 'Information about the data files associated with the experiment.'), ('type', 'array'), ('items', OrderedDict([('title', 'File'), ('description', 'File metadata.'), ('type', 'string'), ('linkTo', 'File')]))])), ('filesets', OrderedDict([('title', 'Sets of Metadata on Data files'), ('description', 'Information about a set of data files associated with the experiment.'), ('type', 'array'), ('exclude_from', ['submit4dn', 'FFedit-create']), ('items', OrderedDict([('title', 'Fileset'), ('description', 'Metadata on a set of files.'), ('type', 'string'), ('linkTo', 'FileSet')]))])), ('quality_metric_flags', OrderedDict([('title', 'Quality Metric Flags'), ('description', 'Flags and associated info for flagging experiments with possible issues'), ('type', 'array'), ('exclude_from', ['submit4dn', 'FFedit-create']), ('items', OrderedDict([('title', 'Flag'), ('description', 'A flag for this experiment with link to associated QC info.'), ('type', 'string'), ('linkTo', 'QualityMetricFlag')]))])), ('follows_sop', OrderedDict([('title', 'Follows SOP'), ('description', 'Flag to indicate if the 4DN SOP for the selected type of experiment was followed - Yes or No'), ('type', 'string'), ('enum', ['Yes', 'No'])]))])), ('facets', OrderedDict([('award.project', OrderedDict([('title', 'Project')])), ('biosample.biosource.individual.organism.name', OrderedDict([('title', 'Organism')])), ('biosample.biosource.biosource_type', OrderedDict([('title', 'Biosource type')])), ('biosample.biosource_summary', OrderedDict([('title', 'Biosource')])), ('biosample.modifications_summary', OrderedDict([('title', 'Modifications')])), ('biosample.treatments_summary', OrderedDict([('title', 'Treatments')])), ('lab.title', OrderedDict([('title', 'Lab')]))]))])</em><a class="headerlink" href="#encoded.types.experiment.Experiment.schema" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="encoded.types.experiment.ExperimentCaptureC">
<em class="property">class </em><code class="descclassname">encoded.types.experiment.</code><code class="descname">ExperimentCaptureC</code><span class="sig-paren">(</span><em>registry</em>, <em>models</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#ExperimentCaptureC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.ExperimentCaptureC" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#encoded.types.experiment.Experiment" title="encoded.types.experiment.Experiment"><code class="xref py py-class docutils literal"><span class="pre">encoded.types.experiment.Experiment</span></code></a></p>
<p>The experiment class for Capture Hi-C experiments.</p>
<dl class="method">
<dt id="encoded.types.experiment.ExperimentCaptureC.display_title">
<code class="descname">display_title</code><span class="sig-paren">(</span><em>request</em>, <em>experiment_type='Undefined'</em>, <em>digestion_enzyme=None</em>, <em>biosample=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#ExperimentCaptureC.display_title"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.ExperimentCaptureC.display_title" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentCaptureC.embedded">
<code class="descname">embedded</code><em class="property"> = ['protocol', 'protocol_variation', 'lab', 'award', 'biosample', 'biosample.biosource', 'biosample.modifications', 'biosample.treatments', 'biosample.biosource.individual.organism', 'digestion_enzyme', 'submitted_by', 'targeted_regions', 'targeted_regions.target', 'targeted_regions.oligo_file']</em><a class="headerlink" href="#encoded.types.experiment.ExperimentCaptureC.embedded" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="encoded.types.experiment.ExperimentCaptureC.experiment_summary">
<code class="descname">experiment_summary</code><span class="sig-paren">(</span><em>request</em>, <em>experiment_type='Undefined'</em>, <em>digestion_enzyme=None</em>, <em>biosample=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#ExperimentCaptureC.experiment_summary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.ExperimentCaptureC.experiment_summary" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentCaptureC.item_type">
<code class="descname">item_type</code><em class="property"> = 'experiment_capture_c'</em><a class="headerlink" href="#encoded.types.experiment.ExperimentCaptureC.item_type" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentCaptureC.name_key">
<code class="descname">name_key</code><em class="property"> = 'accession'</em><a class="headerlink" href="#encoded.types.experiment.ExperimentCaptureC.name_key" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentCaptureC.schema">
<code class="descname">schema</code><em class="property"> = OrderedDict([('title', 'Capture Hi-C Experiment Variables'), ('description', 'Experimental variables for a Capture Hi-C experiment.'), ('id', '/profiles/experiment_capture_c.json'), ('$schema', 'http://json-schema.org/draft-04/schema#'), ('type', 'object'), ('required', ['experiment_type', 'award', 'lab', 'biosample']), ('identifyingProperties', ['uuid', 'accession', 'aliases']), ('additionalProperties', False), ('mixinProperties', [OrderedDict([('$ref', 'mixins.json#/schema_version')]), OrderedDict([('$ref', 'mixins.json#/accession')]), OrderedDict([('$ref', 'mixins.json#/uuid')]), OrderedDict([('$ref', 'mixins.json#/aliases')]), OrderedDict([('$ref', 'mixins.json#/attribution')]), OrderedDict([('$ref', 'mixins.json#/submitted')]), OrderedDict([('$ref', 'mixins.json#/notes')]), OrderedDict([('$ref', 'mixins.json#/references')]), OrderedDict([('$ref', 'mixins.json#/dbxrefs')]), OrderedDict([('$ref', 'mixins.json#/documents')]), OrderedDict([('$ref', 'mixins.json#/library')]), OrderedDict([('$ref', 'mixins.json#/sop_mapping')]), OrderedDict([('$ref', 'experiment.json#/properties')]), OrderedDict([('$ref', 'experiment_hi_c.json#/properties')])]), ('dependencies', OrderedDict([('crosslinking_temperature', ['crosslinking_method', 'crosslinking_time']), ('crosslinking_time', ['crosslinking_method', 'crosslinking_temperature']), ('digestion_temperature', ['digestion_enzyme', 'digestion_time']), ('digestion_time', ['digestion_enzyme', 'digestion_temperature'])])), ('properties', OrderedDict([('schema_version', {'default': '1', 'title': 'Schema Version', 'internal_comment': 'Do not submit, value is assigned by the server. The version of the JSON schema that the server uses to validate the object. Schema version indicates generation of schema used to save version to to enable upgrade steps to work. Individual schemas should set the default.', 'type': 'string', 'exclude_from': ['submit4dn', 'FFedit-create'], 'pattern': '^\\d+(\\.\\d+)*$', 'requestMethod': []}), ('experiment_type', {'title': 'Experiment Type', 'enum': ['capture Hi-C'], 'type': 'string', 'description': 'A controlled term specifying the type of experiment.', 'internal_comment': 'should be ontology term?'}), ('crosslinking_method', {'title': 'Crosslinking Method', 'description': 'Term used for the method for crosslinking chromatin', 'type': 'string', 'internal_comment': 'We might need to change the structure to accomodate user additions', 'enum': ['none', '1% Formaldehyde', '1.3% Formaldehyde']}), ('crosslinking_time', {'title': 'Crosslinking Time (min)', 'description': 'Time of crosslinking step in minutes', 'type': 'number'}), ('crosslinking_temperature', {'title': 'Crosslinking Temperature (°C)', 'description': 'Temperature of crosslinking step in degrees Celsius', 'type': 'number'}), ('digestion_enzyme', {'title': 'Digestion Enzyme', 'description': 'The enzyme used for digestion of the DNA.', 'comment': 'See Enzymes sheet or collection for existing items.', 'type': 'string', 'linkTo': 'Enzyme'}), ('enzyme_lot_number', {'title': 'Enzyme Lot Number', 'description': 'Lot number of batch of enzyme used to digest DNA', 'type': 'string'}), ('digestion_time', {'title': 'Digestion Time (min)', 'description': 'Time of digestion step in minutes', 'type': 'number'}), ('digestion_temperature', {'title': 'Digestion Temperature (°C)', 'description': 'Temperature of digestion step in degrees Celsius', 'type': 'number'}), ('tagging_method', {'title': 'Tagging Method', 'description': 'Information on the biotinylated base used or other tagging info', 'type': 'string', 'internal_comment': 'should this be a controlled CV?'}), ('ligation_time', {'title': 'Ligation Time (min)', 'description': 'Time of ligation step in minutes', 'type': 'number'}), ('ligation_temperature', {'title': 'Ligation Temperature (°C)', 'description': 'Temperature of ligation step in degrees Celsius', 'type': 'number'}), ('ligation_volume', {'title': 'Ligation Volume (ml)', 'description': 'Volume of ligation step in milliliters', 'type': 'number'}), ('biotin_removed', {'title': 'Biotin Removal Step', 'description': 'The optional biotin removal step was performed', 'type': 'string', 'enum': ['Yes', 'No']}), ('status', {'title': 'Status', 'type': 'string', 'default': 'in review by lab', 'permission': 'import_items', 'enum': ['released', 'current', 'revoked', 'deleted', 'replaced', 'in review by lab', 'in review by project', 'released to project']}), ('accession', {'accessionType': 'EX', 'title': 'Accession', 'description': 'A unique identifier to be used to reference the object.', 'internal_comment': 'Only admins are allowed to set or update this value.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'type': 'string', 'format': 'accession', 'permission': 'import_items', 'serverDefault': 'accession'}), ('description', {'title': 'description', 'description': 'A short description of the experiment', 'type': 'string'}), ('protocol', {'title': 'Reference Protocol', 'description': 'Reference Protocol Document.', 'type': 'string', 'linkTo': 'Protocol'}), ('protocol_variation', {'title': 'Protocol Documents', 'description': 'Protocol Documents Describing the variations in the protocol from reference.', 'type': 'array', 'items': {'title': 'Document', 'description': 'A document associated with the specific experimental protocol.', 'type': 'string', 'linkTo': 'Protocol'}}), ('experiment_relation', {'title': 'Experiment relations', 'description': 'All related experiments', 'type': 'array', 'items': {'title': 'Experiment relation', 'type': 'object', 'additionalProperties': False, 'properties': {'relationship_type': {'type': 'string', 'description': 'A controlled term specifying the relationship between experiments.', 'title': 'Relationship Type', 'enum': ['controlled by', 'control for', 'derived from', 'source for']}, 'experiment': {'type': 'string', 'description': 'The related experiment', 'linkTo': 'Experiment'}}}}), ('biosample', {'title': 'Biological Sample', 'description': 'Starting biological material.', 'type': 'string', 'linkTo': 'Biosample'}), ('biosample_quantity', {'title': 'Biological Sample Quantity', 'description': 'The amount of starting Biological sample going into the experiment', 'type': 'number'}), ('biosample_quantity_units', {'title': 'Quantity Units', 'description': 'The units that go along with the biological sample quantity', 'type': 'string', 'enum': ['g', 'mg', 'μg', 'ml', 'cells']}), ('files', {'title': 'Data files', 'description': 'Information about the data files associated with the experiment.', 'type': 'array', 'items': {'title': 'File', 'description': 'File metadata.', 'type': 'string', 'linkTo': 'File'}}), ('filesets', {'title': 'Sets of Metadata on Data files', 'description': 'Information about a set of data files associated with the experiment.', 'type': 'array', 'exclude_from': ['submit4dn', 'FFedit-create'], 'items': {'title': 'Fileset', 'description': 'Metadata on a set of files.', 'type': 'string', 'linkTo': 'FileSet'}}), ('quality_metric_flags', {'title': 'Quality Metric Flags', 'description': 'Flags and associated info for flagging experiments with possible issues', 'type': 'array', 'exclude_from': ['submit4dn', 'FFedit-create'], 'items': {'title': 'Flag', 'description': 'A flag for this experiment with link to associated QC info.', 'type': 'string', 'linkTo': 'QualityMetricFlag'}}), ('follows_sop', {'title': 'Follows SOP', 'description': 'Flag to indicate if the 4DN SOP for the selected type of experiment was followed - Yes or No', 'type': 'string', 'enum': ['Yes', 'No']}), ('sop_mapping', {'title': 'SOP mappping', 'permission': 'import-items', 'exclude_from': 'submit4dn', 'description': 'Whether there is an SOP map and if so which one for fields default values from SOP current at time of submission', 'type': 'object', 'additionalProperties': False, 'properties': {'has_sop': {'title': 'has SOP', 'description': 'does this item have an sop when it was submitted', 'type': 'string', 'enum': ['Yes', 'No']}, 'sop_map': {'title': 'SOP MAP', 'description': 'the SOP map that the item maps to', 'type': 'string', 'linkTo': 'SopMap'}}}), ('average_fragment_size', {'title': 'Average Fragment Size (bp)', 'description': 'Size of average fragment in base pairs after fragmentation step', 'type': 'integer'}), ('fragment_size_range', {'title': 'Fragment size range', 'description': 'The measured size range of the purified nucleic acid, in bp.', 'type': 'string', 'pattern': '(^[0-9]+-[0-9]+$|^[&lt;&gt;][0-9]+$)'}), ('fragmentation_method', {'title': 'Fragmentation method', 'description': 'A short description or reference of the nucleic acid fragmentation protocol used in library preparation, if applicable.', 'type': 'string', 'enum': ['chemical', 'shearing', 'sonication', 'see protocol document', 'none', 'n/a']}), ('library_preparation_date', {'title': 'Library Preparation Date', 'description': 'The date that the library was prepared - submitted as YYYY-MM-DD.', 'type': 'string', 'format': 'date'}), ('fragment_size_selection_method', {'title': 'Size selection method', 'description': 'A short description or reference of the size selection protocol used in library preparation, if applicable.', 'type': 'string', 'enum': ['gel', 'see document', 'SPRI beads']}), ('documents', {'title': 'Documents', 'description': 'Documents that provide additional information (not data file).', 'comment': 'See Documents sheet or collection for existing items.', 'type': 'array', 'default': [], 'uniqueItems': True, 'items': {'title': 'Document', 'description': 'A document that provides additional information (not data file).', 'type': 'string', 'linkTo': 'Document'}}), ('dbxrefs', {'&#64;type': '&#64;id', 'rdfs:subPropertyOf': 'rdfs:seeAlso', 'title': 'External identifiers', 'comment': 'Enter as a database name:identifier eg. HGNC:PARK2', 'description': 'Unique identifiers from external resources.', 'type': 'array', 'default': [], 'clear_create': 'clear', 'uniqueItems': True, 'items': {'title': 'External identifier', 'description': 'A unique identifier from external resource.', 'type': 'string'}}), ('references', {'title': 'References', 'description': 'The publications that provide more information about the object.', 'type': 'array', 'default': [], 'uniqueItems': True, 'items': {'title': 'Reference', 'description': 'A publication that provide smore information about the object.', 'type': 'string', 'linkTo': 'Publication'}}), ('notes', {'title': 'Notes', 'description': 'DCIC internal notes.', 'type': 'string', 'exclude_from': ['submit4dn', 'FFedit-create'], 'elasticsearch_mapping_index_type': {'title': 'Field mapping index type', 'description': 'Defines one of three types of indexing available', 'type': 'string', 'default': 'analyzed', 'enum': ['analyzed', 'not_analyzed', 'no']}}), ('date_created', {'rdfs:subPropertyOf': 'dc:created', 'title': 'Date created', 'exclude_from': 'submit4dn', 'type': 'string', 'anyOf': [{'format': 'date-time'}, {'format': 'date'}], 'serverDefault': 'now', 'permission': 'import_items'}), ('submitted_by', {'rdfs:subPropertyOf': 'dc:creator', 'title': 'Submitted by', 'exclude_from': 'submit4dn', 'type': 'string', 'linkTo': 'User', 'serverDefault': 'userid', 'permission': 'import_items'}), ('lab', {'title': 'Lab', 'description': 'Lab associated with the submission.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'type': 'string', 'linkTo': 'Lab', 'linkSubmitsFor': True, 'default': ''}), ('award', {'title': 'Grant', 'description': 'Grant associated with the submission.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'default': '', 'type': 'string', 'linkTo': 'Award'}), ('aliases', {'title': 'Lab aliases', 'description': 'Lab specific identifiers to reference an object.', 'type': 'array', 'comment': 'Format is colon separated lab name and lab identifier. (e.g. dcic-lab:42).', 'default': [], 'uniqueItems': True, 'clear_create': 'clear', 'items': {'uniqueKey': 'alias', 'title': 'Lab alias', 'description': 'A lab specific identifier to reference an object.', 'type': 'string', 'pattern': '^\\S+:\\S+'}}), ('uuid', {'title': 'UUID', 'type': 'string', 'format': 'uuid', 'exclude_from': ['submit4dn', 'FFedit-create'], 'serverDefault': 'uuid4', 'permission': 'import_items', 'requestMethod': 'POST'}), ('alternate_accessions', {'title': 'Alternate accessions', 'description': 'Accessions previously assigned to objects that have been merged with this object.', 'type': 'array', 'internal_comment': 'Only admins are allowed to set or update this value.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'default': [], 'items': {'title': 'Alternate Accession', 'description': 'An accession previously assigned to an object that has been merged with this object.', 'type': 'string', 'permission': 'import_items', 'format': 'accession'}}), ('rna_tag', {'title': 'Biotinylated RNA base', 'description': 'The biotinylated base used to tag the in vitro transcribed capture RNA', 'type': 'string', 'internal_comment': 'should this be a controlled CV with enum?'}), ('targeted_regions', {'title': 'Targeted Regions', 'description': 'Information about the regions targeted for capture', 'type': 'array', 'items': OrderedDict([('description', 'A region targeted for capture.'), ('title', 'Target Region'), ('type', 'object'), ('additionalProperties', False), ('properties', OrderedDict([('target', OrderedDict([('title', 'Target'), ('description', 'The genes or region that is specifically targeted'), ('type', 'string'), ('linkTo', 'Target')])), ('oligo_file', OrderedDict([('title', 'Oligonucleotide Fasta File'), ('description', 'The multi fasta file that lists all the oligos used for this region'), ('type', 'string'), ('linkTo', 'File')]))]))])})])), ('facets', OrderedDict([('award.project', OrderedDict([('title', 'Project')])), ('biosample.biosource.individual.organism.name', OrderedDict([('title', 'Organism')])), ('biosample.biosource.biosource_type', OrderedDict([('title', 'Biosource type')])), ('biosample.biosource_summary', OrderedDict([('title', 'Biosource')])), ('digestion_enzyme.name', OrderedDict([('title', 'Enzyme')])), ('biosample.modifications_summary', OrderedDict([('title', 'Modifications')])), ('biosample.treatments_summary', OrderedDict([('title', 'Treatments')])), ('lab.title', OrderedDict([('title', 'Lab')]))]))])</em><a class="headerlink" href="#encoded.types.experiment.ExperimentCaptureC.schema" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="encoded.types.experiment.ExperimentHiC">
<em class="property">class </em><code class="descclassname">encoded.types.experiment.</code><code class="descname">ExperimentHiC</code><span class="sig-paren">(</span><em>registry</em>, <em>models</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#ExperimentHiC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.ExperimentHiC" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#encoded.types.experiment.Experiment" title="encoded.types.experiment.Experiment"><code class="xref py py-class docutils literal"><span class="pre">encoded.types.experiment.Experiment</span></code></a></p>
<p>The experiment class for Hi-C experiments.</p>
<dl class="method">
<dt id="encoded.types.experiment.ExperimentHiC.display_title">
<code class="descname">display_title</code><span class="sig-paren">(</span><em>request</em>, <em>experiment_type='Undefined'</em>, <em>digestion_enzyme=None</em>, <em>biosample=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#ExperimentHiC.display_title"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.ExperimentHiC.display_title" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentHiC.embedded">
<code class="descname">embedded</code><em class="property"> = ['protocol', 'protocol_variation', 'lab', 'award', 'biosample', 'biosample.biosource', 'biosample.modifications', 'biosample.treatments', 'biosample.biosource.individual.organism', 'digestion_enzyme', 'submitted_by']</em><a class="headerlink" href="#encoded.types.experiment.ExperimentHiC.embedded" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="encoded.types.experiment.ExperimentHiC.experiment_summary">
<code class="descname">experiment_summary</code><span class="sig-paren">(</span><em>request</em>, <em>experiment_type='Undefined'</em>, <em>digestion_enzyme=None</em>, <em>biosample=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#ExperimentHiC.experiment_summary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.ExperimentHiC.experiment_summary" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentHiC.item_type">
<code class="descname">item_type</code><em class="property"> = 'experiment_hi_c'</em><a class="headerlink" href="#encoded.types.experiment.ExperimentHiC.item_type" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentHiC.name_key">
<code class="descname">name_key</code><em class="property"> = 'accession'</em><a class="headerlink" href="#encoded.types.experiment.ExperimentHiC.name_key" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentHiC.schema">
<code class="descname">schema</code><em class="property"> = OrderedDict([('title', 'Hi-C Experiment Variables'), ('description', 'Experimental variables specific for a Hi-C experiment.'), ('id', '/profiles/experiment_hi_c.json'), ('$schema', 'http://json-schema.org/draft-04/schema#'), ('type', 'object'), ('required', ['experiment_type', 'award', 'lab', 'biosample']), ('identifyingProperties', ['uuid', 'accession', 'aliases']), ('additionalProperties', False), ('mixinProperties', [OrderedDict([('$ref', 'mixins.json#/schema_version')]), OrderedDict([('$ref', 'mixins.json#/accession')]), OrderedDict([('$ref', 'mixins.json#/uuid')]), OrderedDict([('$ref', 'mixins.json#/aliases')]), OrderedDict([('$ref', 'mixins.json#/attribution')]), OrderedDict([('$ref', 'mixins.json#/submitted')]), OrderedDict([('$ref', 'mixins.json#/notes')]), OrderedDict([('$ref', 'mixins.json#/references')]), OrderedDict([('$ref', 'mixins.json#/dbxrefs')]), OrderedDict([('$ref', 'mixins.json#/documents')]), OrderedDict([('$ref', 'mixins.json#/library')]), OrderedDict([('$ref', 'mixins.json#/sop_mapping')]), OrderedDict([('$ref', 'experiment.json#/properties')])]), ('dependencies', OrderedDict([('crosslinking_temperature', ['crosslinking_method', 'crosslinking_time']), ('crosslinking_time', ['crosslinking_method', 'crosslinking_temperature']), ('digestion_temperature', ['digestion_enzyme', 'digestion_time']), ('digestion_time', ['digestion_enzyme', 'digestion_temperature'])])), ('properties', OrderedDict([('status', {'title': 'Status', 'type': 'string', 'default': 'in review by lab', 'permission': 'import_items', 'enum': ['released', 'current', 'revoked', 'deleted', 'replaced', 'in review by lab', 'in review by project', 'released to project']}), ('accession', {'accessionType': 'EX', 'title': 'Accession', 'description': 'A unique identifier to be used to reference the object.', 'internal_comment': 'Only admins are allowed to set or update this value.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'type': 'string', 'format': 'accession', 'permission': 'import_items', 'serverDefault': 'accession'}), ('description', {'title': 'description', 'description': 'A short description of the experiment', 'type': 'string'}), ('experiment_type', {'type': 'string', 'description': 'A controlled term specifying the type of experiment.', 'internal_comment': 'should be ontology term?', 'title': 'Experiment Type', 'enum': ['in situ Hi-C', 'dilution Hi-C', 'micro-C', 'DNase Hi-C']}), ('protocol', {'title': 'Reference Protocol', 'description': 'Reference Protocol Document.', 'type': 'string', 'linkTo': 'Protocol'}), ('protocol_variation', {'title': 'Protocol Documents', 'description': 'Protocol Documents Describing the variations in the protocol from reference.', 'type': 'array', 'items': {'title': 'Document', 'description': 'A document associated with the specific experimental protocol.', 'type': 'string', 'linkTo': 'Protocol'}}), ('experiment_relation', {'title': 'Experiment relations', 'description': 'All related experiments', 'type': 'array', 'items': {'title': 'Experiment relation', 'type': 'object', 'additionalProperties': False, 'properties': {'relationship_type': {'type': 'string', 'description': 'A controlled term specifying the relationship between experiments.', 'title': 'Relationship Type', 'enum': ['controlled by', 'control for', 'derived from', 'source for']}, 'experiment': {'type': 'string', 'description': 'The related experiment', 'linkTo': 'Experiment'}}}}), ('biosample', {'title': 'Biological Sample', 'description': 'Starting biological material.', 'type': 'string', 'linkTo': 'Biosample'}), ('biosample_quantity', {'title': 'Biological Sample Quantity', 'description': 'The amount of starting Biological sample going into the experiment', 'type': 'number'}), ('biosample_quantity_units', {'title': 'Quantity Units', 'description': 'The units that go along with the biological sample quantity', 'type': 'string', 'enum': ['g', 'mg', 'μg', 'ml', 'cells']}), ('files', {'title': 'Data files', 'description': 'Information about the data files associated with the experiment.', 'type': 'array', 'items': {'title': 'File', 'description': 'File metadata.', 'type': 'string', 'linkTo': 'File'}}), ('filesets', {'title': 'Sets of Metadata on Data files', 'description': 'Information about a set of data files associated with the experiment.', 'type': 'array', 'exclude_from': ['submit4dn', 'FFedit-create'], 'items': {'title': 'Fileset', 'description': 'Metadata on a set of files.', 'type': 'string', 'linkTo': 'FileSet'}}), ('quality_metric_flags', {'title': 'Quality Metric Flags', 'description': 'Flags and associated info for flagging experiments with possible issues', 'type': 'array', 'exclude_from': ['submit4dn', 'FFedit-create'], 'items': {'title': 'Flag', 'description': 'A flag for this experiment with link to associated QC info.', 'type': 'string', 'linkTo': 'QualityMetricFlag'}}), ('follows_sop', {'title': 'Follows SOP', 'description': 'Flag to indicate if the 4DN SOP for the selected type of experiment was followed - Yes or No', 'type': 'string', 'enum': ['Yes', 'No']}), ('sop_mapping', {'title': 'SOP mappping', 'permission': 'import-items', 'exclude_from': 'submit4dn', 'description': 'Whether there is an SOP map and if so which one for fields default values from SOP current at time of submission', 'type': 'object', 'additionalProperties': False, 'properties': {'has_sop': {'title': 'has SOP', 'description': 'does this item have an sop when it was submitted', 'type': 'string', 'enum': ['Yes', 'No']}, 'sop_map': {'title': 'SOP MAP', 'description': 'the SOP map that the item maps to', 'type': 'string', 'linkTo': 'SopMap'}}}), ('average_fragment_size', {'title': 'Average Fragment Size (bp)', 'description': 'Size of average fragment in base pairs after fragmentation step', 'type': 'integer'}), ('fragment_size_range', {'title': 'Fragment size range', 'description': 'The measured size range of the purified nucleic acid, in bp.', 'type': 'string', 'pattern': '(^[0-9]+-[0-9]+$|^[&lt;&gt;][0-9]+$)'}), ('fragmentation_method', {'title': 'Fragmentation method', 'description': 'A short description or reference of the nucleic acid fragmentation protocol used in library preparation, if applicable.', 'type': 'string', 'enum': ['chemical', 'shearing', 'sonication', 'see protocol document', 'none', 'n/a']}), ('library_preparation_date', {'title': 'Library Preparation Date', 'description': 'The date that the library was prepared - submitted as YYYY-MM-DD.', 'type': 'string', 'format': 'date'}), ('fragment_size_selection_method', {'title': 'Size selection method', 'description': 'A short description or reference of the size selection protocol used in library preparation, if applicable.', 'type': 'string', 'enum': ['gel', 'see document', 'SPRI beads']}), ('documents', {'title': 'Documents', 'description': 'Documents that provide additional information (not data file).', 'comment': 'See Documents sheet or collection for existing items.', 'type': 'array', 'default': [], 'uniqueItems': True, 'items': {'title': 'Document', 'description': 'A document that provides additional information (not data file).', 'type': 'string', 'linkTo': 'Document'}}), ('dbxrefs', {'&#64;type': '&#64;id', 'rdfs:subPropertyOf': 'rdfs:seeAlso', 'title': 'External identifiers', 'comment': 'Enter as a database name:identifier eg. HGNC:PARK2', 'description': 'Unique identifiers from external resources.', 'type': 'array', 'default': [], 'clear_create': 'clear', 'uniqueItems': True, 'items': {'title': 'External identifier', 'description': 'A unique identifier from external resource.', 'type': 'string'}}), ('references', {'title': 'References', 'description': 'The publications that provide more information about the object.', 'type': 'array', 'default': [], 'uniqueItems': True, 'items': {'title': 'Reference', 'description': 'A publication that provide smore information about the object.', 'type': 'string', 'linkTo': 'Publication'}}), ('notes', {'title': 'Notes', 'description': 'DCIC internal notes.', 'type': 'string', 'exclude_from': ['submit4dn', 'FFedit-create'], 'elasticsearch_mapping_index_type': {'title': 'Field mapping index type', 'description': 'Defines one of three types of indexing available', 'type': 'string', 'default': 'analyzed', 'enum': ['analyzed', 'not_analyzed', 'no']}}), ('date_created', {'rdfs:subPropertyOf': 'dc:created', 'title': 'Date created', 'exclude_from': 'submit4dn', 'type': 'string', 'anyOf': [{'format': 'date-time'}, {'format': 'date'}], 'serverDefault': 'now', 'permission': 'import_items'}), ('submitted_by', {'rdfs:subPropertyOf': 'dc:creator', 'title': 'Submitted by', 'exclude_from': 'submit4dn', 'type': 'string', 'linkTo': 'User', 'serverDefault': 'userid', 'permission': 'import_items'}), ('lab', {'title': 'Lab', 'description': 'Lab associated with the submission.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'type': 'string', 'linkTo': 'Lab', 'linkSubmitsFor': True, 'default': ''}), ('award', {'title': 'Grant', 'description': 'Grant associated with the submission.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'default': '', 'type': 'string', 'linkTo': 'Award'}), ('aliases', {'title': 'Lab aliases', 'description': 'Lab specific identifiers to reference an object.', 'type': 'array', 'comment': 'Format is colon separated lab name and lab identifier. (e.g. dcic-lab:42).', 'default': [], 'uniqueItems': True, 'clear_create': 'clear', 'items': {'uniqueKey': 'alias', 'title': 'Lab alias', 'description': 'A lab specific identifier to reference an object.', 'type': 'string', 'pattern': '^\\S+:\\S+'}}), ('uuid', {'title': 'UUID', 'type': 'string', 'format': 'uuid', 'exclude_from': ['submit4dn', 'FFedit-create'], 'serverDefault': 'uuid4', 'permission': 'import_items', 'requestMethod': 'POST'}), ('alternate_accessions', {'title': 'Alternate accessions', 'description': 'Accessions previously assigned to objects that have been merged with this object.', 'type': 'array', 'internal_comment': 'Only admins are allowed to set or update this value.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'default': [], 'items': {'title': 'Alternate Accession', 'description': 'An accession previously assigned to an object that has been merged with this object.', 'type': 'string', 'permission': 'import_items', 'format': 'accession'}}), ('schema_version', {'title': 'Schema Version', 'internal_comment': 'Do not submit, value is assigned by the server. The version of the JSON schema that the server uses to validate the object. Schema version indicates generation of schema used to save version to to enable upgrade steps to work. Individual schemas should set the default.', 'type': 'string', 'exclude_from': ['submit4dn', 'FFedit-create'], 'pattern': '^\\d+(\\.\\d+)*$', 'requestMethod': [], 'default': '1'}), ('crosslinking_method', {'title': 'Crosslinking Method', 'description': 'Term used for the method for crosslinking chromatin', 'type': 'string', 'internal_comment': 'We might need to change the structure to accomodate user additions', 'enum': ['none', '1% Formaldehyde', '1.3% Formaldehyde']}), ('crosslinking_time', {'title': 'Crosslinking Time (min)', 'description': 'Time of crosslinking step in minutes', 'type': 'number'}), ('crosslinking_temperature', {'title': 'Crosslinking Temperature (°C)', 'description': 'Temperature of crosslinking step in degrees Celsius', 'type': 'number'}), ('digestion_enzyme', {'title': 'Digestion Enzyme', 'description': 'The enzyme used for digestion of the DNA.', 'comment': 'See Enzymes sheet or collection for existing items.', 'type': 'string', 'linkTo': 'Enzyme'}), ('enzyme_lot_number', {'title': 'Enzyme Lot Number', 'description': 'Lot number of batch of enzyme used to digest DNA', 'type': 'string'}), ('digestion_time', {'title': 'Digestion Time (min)', 'description': 'Time of digestion step in minutes', 'type': 'number'}), ('digestion_temperature', {'title': 'Digestion Temperature (°C)', 'description': 'Temperature of digestion step in degrees Celsius', 'type': 'number'}), ('tagging_method', {'title': 'Tagging Method', 'description': 'Information on the biotinylated base used or other tagging info', 'type': 'string', 'internal_comment': 'should this be a controlled CV?'}), ('ligation_time', {'title': 'Ligation Time (min)', 'description': 'Time of ligation step in minutes', 'type': 'number'}), ('ligation_temperature', {'title': 'Ligation Temperature (°C)', 'description': 'Temperature of ligation step in degrees Celsius', 'type': 'number'}), ('ligation_volume', {'title': 'Ligation Volume (ml)', 'description': 'Volume of ligation step in milliliters', 'type': 'number'}), ('biotin_removed', {'title': 'Biotin Removal Step', 'description': 'The optional biotin removal step was performed', 'type': 'string', 'enum': ['Yes', 'No']})])), ('facets', OrderedDict([('award.project', OrderedDict([('title', 'Project')])), ('biosample.biosource.individual.organism.name', OrderedDict([('title', 'Organism')])), ('biosample.biosource.biosource_type', OrderedDict([('title', 'Biosource type')])), ('biosample.biosource_summary', OrderedDict([('title', 'Biosource')])), ('digestion_enzyme.name', OrderedDict([('title', 'Enzyme')])), ('biosample.modifications_summary', OrderedDict([('title', 'Modifications')])), ('biosample.treatments_summary', OrderedDict([('title', 'Treatments')])), ('lab.title', OrderedDict([('title', 'Lab')]))]))])</em><a class="headerlink" href="#encoded.types.experiment.ExperimentHiC.schema" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="encoded.types.experiment.ExperimentRepliseq">
<em class="property">class </em><code class="descclassname">encoded.types.experiment.</code><code class="descname">ExperimentRepliseq</code><span class="sig-paren">(</span><em>registry</em>, <em>models</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#ExperimentRepliseq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.ExperimentRepliseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#encoded.types.experiment.Experiment" title="encoded.types.experiment.Experiment"><code class="xref py py-class docutils literal"><span class="pre">encoded.types.experiment.Experiment</span></code></a></p>
<p>The experiment class for Repliseq experiments.</p>
<dl class="method">
<dt id="encoded.types.experiment.ExperimentRepliseq.display_title">
<code class="descname">display_title</code><span class="sig-paren">(</span><em>request</em>, <em>experiment_type='Undefined'</em>, <em>cell_cycle_stage=None</em>, <em>biosample=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#ExperimentRepliseq.display_title"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.ExperimentRepliseq.display_title" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentRepliseq.embedded">
<code class="descname">embedded</code><em class="property"> = ['protocol', 'protocol_variation', 'lab', 'award', 'biosample', 'biosample.biosource', 'biosample.modifications', 'biosample.treatments', 'biosample.biosource.individual.organism', 'submitted_by']</em><a class="headerlink" href="#encoded.types.experiment.ExperimentRepliseq.embedded" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="encoded.types.experiment.ExperimentRepliseq.experiment_summary">
<code class="descname">experiment_summary</code><span class="sig-paren">(</span><em>request</em>, <em>experiment_type='Undefined'</em>, <em>cell_cycle_stage=None</em>, <em>biosample=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/encoded/types/experiment.html#ExperimentRepliseq.experiment_summary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#encoded.types.experiment.ExperimentRepliseq.experiment_summary" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentRepliseq.item_type">
<code class="descname">item_type</code><em class="property"> = 'experiment_repliseq'</em><a class="headerlink" href="#encoded.types.experiment.ExperimentRepliseq.item_type" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentRepliseq.name_key">
<code class="descname">name_key</code><em class="property"> = 'accession'</em><a class="headerlink" href="#encoded.types.experiment.ExperimentRepliseq.name_key" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="encoded.types.experiment.ExperimentRepliseq.schema">
<code class="descname">schema</code><em class="property"> = OrderedDict([('title', 'Repliseq Experiment Variables'), ('description', 'Experimental variables for a Repliseq experiment.'), ('id', '/profiles/experiment_repliseq.json'), ('$schema', 'http://json-schema.org/draft-04/schema#'), ('type', 'object'), ('required', ['experiment_type', 'award', 'lab', 'biosample']), ('identifyingProperties', ['uuid', 'accession', 'aliases']), ('additionalProperties', False), ('mixinProperties', [OrderedDict([('$ref', 'mixins.json#/schema_version')]), OrderedDict([('$ref', 'mixins.json#/accession')]), OrderedDict([('$ref', 'mixins.json#/uuid')]), OrderedDict([('$ref', 'mixins.json#/aliases')]), OrderedDict([('$ref', 'mixins.json#/attribution')]), OrderedDict([('$ref', 'mixins.json#/submitted')]), OrderedDict([('$ref', 'mixins.json#/notes')]), OrderedDict([('$ref', 'mixins.json#/references')]), OrderedDict([('$ref', 'mixins.json#/dbxrefs')]), OrderedDict([('$ref', 'mixins.json#/documents')]), OrderedDict([('$ref', 'mixins.json#/library')]), OrderedDict([('$ref', 'mixins.json#/sop_mapping')]), OrderedDict([('$ref', 'experiment.json#/properties')])]), ('properties', OrderedDict([('status', {'title': 'Status', 'type': 'string', 'default': 'in review by lab', 'permission': 'import_items', 'enum': ['released', 'current', 'revoked', 'deleted', 'replaced', 'in review by lab', 'in review by project', 'released to project']}), ('accession', {'accessionType': 'EX', 'title': 'Accession', 'description': 'A unique identifier to be used to reference the object.', 'internal_comment': 'Only admins are allowed to set or update this value.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'type': 'string', 'format': 'accession', 'permission': 'import_items', 'serverDefault': 'accession'}), ('description', {'title': 'description', 'description': 'A short description of the experiment', 'type': 'string'}), ('experiment_type', {'type': 'string', 'description': 'A controlled term specifying the type of experiment.', 'internal_comment': 'should be ontology term?', 'enum': ['repliseq']}), ('protocol', {'title': 'Reference Protocol', 'description': 'Reference Protocol Document.', 'type': 'string', 'linkTo': 'Protocol'}), ('protocol_variation', {'title': 'Protocol Documents', 'description': 'Protocol Documents Describing the variations in the protocol from reference.', 'type': 'array', 'items': {'title': 'Document', 'description': 'A document associated with the specific experimental protocol.', 'type': 'string', 'linkTo': 'Protocol'}}), ('experiment_relation', {'title': 'Experiment relations', 'description': 'All related experiments', 'type': 'array', 'items': {'title': 'Experiment relation', 'type': 'object', 'additionalProperties': False, 'properties': {'relationship_type': {'type': 'string', 'description': 'A controlled term specifying the relationship between experiments.', 'title': 'Relationship Type', 'enum': ['controlled by', 'control for', 'derived from', 'source for']}, 'experiment': {'type': 'string', 'description': 'The related experiment', 'linkTo': 'Experiment'}}}}), ('biosample', {'title': 'Biological Sample', 'description': 'Starting biological material.', 'type': 'string', 'linkTo': 'Biosample'}), ('biosample_quantity', {'title': 'Biological Sample Quantity', 'description': 'The amount of starting Biological sample going into the experiment', 'type': 'number'}), ('biosample_quantity_units', {'title': 'Quantity Units', 'description': 'The units that go along with the biological sample quantity', 'type': 'string', 'enum': ['g', 'mg', 'μg', 'ml', 'cells']}), ('files', {'title': 'Data files', 'description': 'Information about the data files associated with the experiment.', 'type': 'array', 'items': {'title': 'File', 'description': 'File metadata.', 'type': 'string', 'linkTo': 'File'}}), ('filesets', {'title': 'Sets of Metadata on Data files', 'description': 'Information about a set of data files associated with the experiment.', 'type': 'array', 'exclude_from': ['submit4dn', 'FFedit-create'], 'items': {'title': 'Fileset', 'description': 'Metadata on a set of files.', 'type': 'string', 'linkTo': 'FileSet'}}), ('quality_metric_flags', {'title': 'Quality Metric Flags', 'description': 'Flags and associated info for flagging experiments with possible issues', 'type': 'array', 'exclude_from': ['submit4dn', 'FFedit-create'], 'items': {'title': 'Flag', 'description': 'A flag for this experiment with link to associated QC info.', 'type': 'string', 'linkTo': 'QualityMetricFlag'}}), ('follows_sop', {'title': 'Follows SOP', 'description': 'Flag to indicate if the 4DN SOP for the selected type of experiment was followed - Yes or No', 'type': 'string', 'enum': ['Yes', 'No']}), ('sop_mapping', {'title': 'SOP mappping', 'permission': 'import-items', 'exclude_from': 'submit4dn', 'description': 'Whether there is an SOP map and if so which one for fields default values from SOP current at time of submission', 'type': 'object', 'additionalProperties': False, 'properties': {'has_sop': {'title': 'has SOP', 'description': 'does this item have an sop when it was submitted', 'type': 'string', 'enum': ['Yes', 'No']}, 'sop_map': {'title': 'SOP MAP', 'description': 'the SOP map that the item maps to', 'type': 'string', 'linkTo': 'SopMap'}}}), ('average_fragment_size', {'title': 'Average Fragment Size (bp)', 'description': 'Size of average fragment in base pairs after fragmentation step', 'type': 'integer'}), ('fragment_size_range', {'title': 'Fragment size range', 'description': 'The measured size range of the purified nucleic acid, in bp.', 'type': 'string', 'pattern': '(^[0-9]+-[0-9]+$|^[&lt;&gt;][0-9]+$)'}), ('fragmentation_method', {'title': 'Fragmentation method', 'description': 'A short description or reference of the nucleic acid fragmentation protocol used in library preparation, if applicable.', 'type': 'string', 'enum': ['chemical', 'shearing', 'sonication', 'see protocol document', 'none', 'n/a']}), ('library_preparation_date', {'title': 'Library Preparation Date', 'description': 'The date that the library was prepared - submitted as YYYY-MM-DD.', 'type': 'string', 'format': 'date'}), ('fragment_size_selection_method', {'title': 'Size selection method', 'description': 'A short description or reference of the size selection protocol used in library preparation, if applicable.', 'type': 'string', 'enum': ['gel', 'see document', 'SPRI beads']}), ('documents', {'title': 'Documents', 'description': 'Documents that provide additional information (not data file).', 'comment': 'See Documents sheet or collection for existing items.', 'type': 'array', 'default': [], 'uniqueItems': True, 'items': {'title': 'Document', 'description': 'A document that provides additional information (not data file).', 'type': 'string', 'linkTo': 'Document'}}), ('dbxrefs', {'&#64;type': '&#64;id', 'rdfs:subPropertyOf': 'rdfs:seeAlso', 'title': 'External identifiers', 'comment': 'Enter as a database name:identifier eg. HGNC:PARK2', 'description': 'Unique identifiers from external resources.', 'type': 'array', 'default': [], 'clear_create': 'clear', 'uniqueItems': True, 'items': {'title': 'External identifier', 'description': 'A unique identifier from external resource.', 'type': 'string'}}), ('references', {'title': 'References', 'description': 'The publications that provide more information about the object.', 'type': 'array', 'default': [], 'uniqueItems': True, 'items': {'title': 'Reference', 'description': 'A publication that provide smore information about the object.', 'type': 'string', 'linkTo': 'Publication'}}), ('notes', {'title': 'Notes', 'description': 'DCIC internal notes.', 'type': 'string', 'exclude_from': ['submit4dn', 'FFedit-create'], 'elasticsearch_mapping_index_type': {'title': 'Field mapping index type', 'description': 'Defines one of three types of indexing available', 'type': 'string', 'default': 'analyzed', 'enum': ['analyzed', 'not_analyzed', 'no']}}), ('date_created', {'rdfs:subPropertyOf': 'dc:created', 'title': 'Date created', 'exclude_from': 'submit4dn', 'type': 'string', 'anyOf': [{'format': 'date-time'}, {'format': 'date'}], 'serverDefault': 'now', 'permission': 'import_items'}), ('submitted_by', {'rdfs:subPropertyOf': 'dc:creator', 'title': 'Submitted by', 'exclude_from': 'submit4dn', 'type': 'string', 'linkTo': 'User', 'serverDefault': 'userid', 'permission': 'import_items'}), ('lab', {'title': 'Lab', 'description': 'Lab associated with the submission.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'type': 'string', 'linkTo': 'Lab', 'linkSubmitsFor': True, 'default': ''}), ('award', {'title': 'Grant', 'description': 'Grant associated with the submission.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'default': '', 'type': 'string', 'linkTo': 'Award'}), ('aliases', {'title': 'Lab aliases', 'description': 'Lab specific identifiers to reference an object.', 'type': 'array', 'comment': 'Format is colon separated lab name and lab identifier. (e.g. dcic-lab:42).', 'default': [], 'uniqueItems': True, 'clear_create': 'clear', 'items': {'uniqueKey': 'alias', 'title': 'Lab alias', 'description': 'A lab specific identifier to reference an object.', 'type': 'string', 'pattern': '^\\S+:\\S+'}}), ('uuid', {'title': 'UUID', 'type': 'string', 'format': 'uuid', 'exclude_from': ['submit4dn', 'FFedit-create'], 'serverDefault': 'uuid4', 'permission': 'import_items', 'requestMethod': 'POST'}), ('alternate_accessions', {'title': 'Alternate accessions', 'description': 'Accessions previously assigned to objects that have been merged with this object.', 'type': 'array', 'internal_comment': 'Only admins are allowed to set or update this value.', 'exclude_from': ['submit4dn', 'FFedit-create'], 'default': [], 'items': {'title': 'Alternate Accession', 'description': 'An accession previously assigned to an object that has been merged with this object.', 'type': 'string', 'permission': 'import_items', 'format': 'accession'}}), ('schema_version', {'title': 'Schema Version', 'internal_comment': 'Do not submit, value is assigned by the server. The version of the JSON schema that the server uses to validate the object. Schema version indicates generation of schema used to save version to to enable upgrade steps to work. Individual schemas should set the default.', 'type': 'string', 'exclude_from': ['submit4dn', 'FFedit-create'], 'pattern': '^\\d+(\\.\\d+)*$', 'requestMethod': [], 'default': '1'}), ('dna_label', {'title': 'DNA label', 'description': 'The tag or chemical used to label the newly replicated DNA - usually BrdU', 'type': 'string', 'comment': 'e.g. BrdU', 'internal_comment': 'an enum/default of BrdU'}), ('labeling_time', {'title': 'Labeling time', 'description': 'Time in minutes of labeling pulse', 'type': 'number'}), ('antibody', {'title': 'Antibody', 'description': 'ENCODE accession for antibody used to pull down the labeled DNA', 'type': 'string', 'internal_comment': 'Can we depend on encode having antibody or allow other string?', 'pattern': 'ENCAB.+'}), ('antibody_lot_id', {'title': 'Antibody Lot ID', 'description': 'Identifier for the specific lot of antibody used', 'type': 'string'}), ('cell_cycle_phase', {'title': 'Cell Cycle Phase', 'description': 'After sorting the cell cycle phase that the cell fraction falls in', 'type': 'string', 'internal_comment': 'Link with a specific ontology', 'enum': ['G1', 'S', 'G2', 'M']}), ('stage_fraction', {'title': 'Cell Cycle Stage Fraction', 'description': 'A decription of the sub stage prepared after sorting', 'comment': 'Repliseq experiments can be 2-phase, 6-phase, 8-phase ... This field is used to further describe the fraction represented in this experiment', 'type': 'string'}), ('cell_sorting_protocol', {'title': 'Sorting Protocol', 'description': 'Protocol with details on cell sorting by FACS', 'type': 'string', 'linkTo': 'Protocol'})])), ('facets', OrderedDict([('award.project', OrderedDict([('title', 'Project')])), ('biosample.biosource.individual.organism.name', OrderedDict([('title', 'Organism')])), ('biosample.biosource.biosource_type', OrderedDict([('title', 'Biosource type')])), ('biosample.biosource_summary', OrderedDict([('title', 'Biosource')])), ('biosample.modifications_summary', OrderedDict([('title', 'Modifications')])), ('biosample.treatments_summary', OrderedDict([('title', 'Treatments')])), ('lab.title', OrderedDict([('title', 'Lab')]))]))])</em><a class="headerlink" href="#encoded.types.experiment.ExperimentRepliseq.schema" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="encoded.types.biosource.html"
                        title="previous chapter">encoded.types.biosource module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="encoded.types.experiment_set.html"
                        title="next chapter">encoded.types.experiment_set module</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pydoc/encoded.types.experiment.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="encoded.types.experiment_set.html" title="encoded.types.experiment_set module"
             >next</a> |</li>
        <li class="right" >
          <a href="encoded.types.biosource.html" title="encoded.types.biosource module"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">4D Nucleome Portal 1.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="modules.html" >Python API Reference</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="encoded.html" >encoded package</a> &#187;</li>
          <li class="nav-item nav-item-3"><a href="encoded.types.html" >encoded.types package</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, 4DN-DCIC.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.3.
    </div>
  </body>
</html>